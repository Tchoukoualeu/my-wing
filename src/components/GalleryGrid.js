import React, { useState } from 'react';
import { makeStyles } from '@material-ui/core/styles';
import GridList from '@material-ui/core/GridList';
import GridListTile from '@material-ui/core/GridListTile';
import GridListTileBar from '@material-ui/core/GridListTileBar';
import { withWidth, isWidthUp } from '@material-ui/core';
import ImageModal from './ImageModal';

const useStyles = makeStyles((theme) => ({
  root: {
    display: 'flex',
    flexWrap: 'wrap',
    justifyContent: 'space-around',
    overflow: 'hidden',
    backgroundColor: theme.palette.background.paper
  },
  pointer: {
    cursor: 'pointer'
  }
}));

const GalleryGrid = ({ images, width }) => {
  const classes = useStyles();

  const [imageIsOpen, setImageIsOpen] = useState(false);
  const [imageDetails, setImageDetails] = useState(false);

  const handleOpen = (imgDetails) => {
    setImageIsOpen(true);
    setImageDetails(imgDetails);
  };

  const handleClose = () => {
    setImageIsOpen(false);
    setImageDetails(false);
  };

  const getGridListCols = () => {
    if (isWidthUp('lg', width)) {
      return 5;
    }

    if (isWidthUp('md', width)) {
      return 4;
    }

    if (isWidthUp('sm', width)) {
      return 3;
    }

    return 2;
  }

  const getImageParts = (image) => {
    let link = image.link,
        title = image.title,
        imageObj = {
          title,
          desc: image.description,
          link,
          type: image.type,
          ups: image.ups,
          downs: image.downs,
          score: image.score
        };

    if (!!image.type && image.type === 'video/mp4') {
      link = image.gifv.slice(0, -1);
    }

    if (typeof image.images !== 'undefined') {
      const curImage = image.images[0];

      link = curImage.link;
      imageObj.link = curImage.link;
      if (curImage.type === 'video/mp4') {
        link = curImage.gifv.slice(0, -1);
      }
      if (!title) {
        title = curImage.title;
        imageObj.title = curImage.title;
      }
      if (!image.description) {
        imageObj.desc = curImage.description;
      }
      if (!image.type) {
        imageObj.type = curImage.type;
      }
      if (!image.ups) {
        imageObj.ups = curImage.ups;
      }
      if (!image.downs) {
        imageObj.downs = curImage.downs;
      }
      if (!image.score) {
        imageObj.score = curImage.score;
      }
    }

    return [link, title, imageObj];
  };

  return (
    <div className={classes.root}>
      <ImageModal
        imageIsOpen={imageIsOpen}
        imageDetails={imageDetails}
        handleClose={handleClose}
      />
      <GridList
        cols={getGridListCols()}
        spacing={(isWidthUp('md', width) ? 20 : 4)}
      >
        {images.map((image, key) => {
          const [link, title, imageObj] = getImageParts(image);

          return (
            <GridListTile className={classes.pointer} key={key} onClick={() => handleOpen(imageObj)}>
              <img src={link} alt={title} />
              <GridListTileBar title={title} />
            </GridListTile>
          );
        })}
      </GridList>
    </div>
  );
};

export default withWidth()(GalleryGrid);