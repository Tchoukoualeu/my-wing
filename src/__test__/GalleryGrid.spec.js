import { render, screen } from "@testing-library/react";
import userEvent from "@testing-library/user-event";
import React from "react";
import { create } from "react-test-renderer";
import GalleryGrid from "../components/GalleryGrid";

describe("GalleryGrid component", () => {
  test("Matches the snapshots for GalleryGrid with empty images", () => {
    const galleryGrid = create(
      <GalleryGrid
        images={[]}
      />
    );

    expect(galleryGrid.toJSON()).toMatchSnapshot();
  });

  test("Matches the snapshots for GalleryGrid", () => {
    const images = [
      {
        title: 'title',
        description: 'desc',
        type: 'image/png',
        link: 'link',
        ups: 1,
        downs: 2,
        score: 3
      },
      {
        link: 'link',
        images: [
          {
            title: 'title1',
            description: 'desc1',
            type: 'video/mp4',
            link: 'link1_mp4',
            gifv: 'link1.gifv',
            ups: 1,
            downs: 2,
            score: 3
          }
        ]
      }, {
        title: 'title',
        description: 'desc',
        type: 'image/png',
        link: 'link',
        ups: 1,
        downs: 2,
        score: 3,
        images: [
          {
            title: 'title1',
            description: 'desc1',
            type: 'video/mp4',
            link: 'link1_mp4',
            gifv: 'link1.gifv',
            ups: 1,
            downs: 2,
            score: 3
          }
        ]
      },
      {
        title: 'title2',
        description: 'desc2',
        type: 'video/mp4',
        link: 'link2',
        gifv: 'link2.gifv'
      },
      {
        link: 'link',
        description: 'desc3',
        images: [
          {
            title: 'title1',
            type: 'image/jpeg',
            link: 'link1_mp4',
            gifv: 'link1.gifv',
            ups: 1,
            downs: 2,
            score: 3
          }
        ]
      }
    ];
    const galleryGrid = create(
      <GalleryGrid
        images={images}
        width='xs'
      />
    );

    expect(galleryGrid.toJSON()).toMatchSnapshot();
  });

  test("Matches the snapshots for GalleryGrid with different screen width", () => {
    const images = [
      {
        title: 'title',
        description: 'desc',
        type: 'image/png',
        link: 'link'
      }
    ];
    ['xs', 'sm', 'md', 'lg'].forEach(width => {
      const galleryGrid = create(
        <GalleryGrid
          images={images}
          width={width}
        />
      );

      expect(galleryGrid.toJSON()).toMatchSnapshot();
    });
  });

  test("It shows that GalleryGrid dimms and appears", () => {
    const images = [
      {
        title: 'title',
        description: 'desc',
        type: 'image/png',
        link: 'link'
      }
    ];
    const { container } = render(<GalleryGrid images={images} width='sm' />);

    userEvent.click(container.getElementsByTagName('img')[0]);

    expect(container.getElementsByTagName('img')).not.toBe();

    userEvent.click(screen.getByRole('button'));
  });

  test("It shows that GalleryGrid dimms and appears with another image", () => {
    const images = [
      {
        title: 'title',
        description: 'desc',
        type: 'video/mp4',
        link: 'link',
        gifv: 'link.difv',
        ups: 1,
        downs: 2,
        score: 3
      }
    ];
    const { container } = render(<GalleryGrid images={images} width='sm' />);

    userEvent.click(container.getElementsByTagName('img')[0]);

    expect(container.getElementsByTagName('img')).not.toBe();

    userEvent.click(screen.getByRole('button'));
  });
});