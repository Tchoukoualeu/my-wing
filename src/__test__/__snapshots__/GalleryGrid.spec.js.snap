// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`GalleryGrid component Matches the snapshots for GalleryGrid 1`] = `
<div
  className="makeStyles-root-1"
>
  <ul
    className="MuiGridList-root"
    style={
      Object {
        "margin": -2,
      }
    }
  >
    <li
      className="MuiGridListTile-root"
      onClick={[Function]}
      style={
        Object {
          "height": 184,
          "padding": 2,
          "width": "50%",
        }
      }
    >
      <div
        className="MuiGridListTile-tile"
      >
        <img
          alt="title"
          src="link"
        />
        <div
          className="MuiGridListTileBar-root MuiGridListTileBar-titlePositionBottom"
        >
          <div
            className="MuiGridListTileBar-titleWrap"
          >
            <div
              className="MuiGridListTileBar-title"
            >
              title
            </div>
          </div>
        </div>
      </div>
    </li>
    <li
      className="MuiGridListTile-root"
      onClick={[Function]}
      style={
        Object {
          "height": 184,
          "padding": 2,
          "width": "50%",
        }
      }
    >
      <div
        className="MuiGridListTile-tile"
      >
        <img
          alt="title1"
          src="link1.gif"
        />
        <div
          className="MuiGridListTileBar-root MuiGridListTileBar-titlePositionBottom"
        >
          <div
            className="MuiGridListTileBar-titleWrap"
          >
            <div
              className="MuiGridListTileBar-title"
            >
              title1
            </div>
          </div>
        </div>
      </div>
    </li>
    <li
      className="MuiGridListTile-root"
      onClick={[Function]}
      style={
        Object {
          "height": 184,
          "padding": 2,
          "width": "50%",
        }
      }
    >
      <div
        className="MuiGridListTile-tile"
      >
        <img
          alt="title"
          src="link1.gif"
        />
        <div
          className="MuiGridListTileBar-root MuiGridListTileBar-titlePositionBottom"
        >
          <div
            className="MuiGridListTileBar-titleWrap"
          >
            <div
              className="MuiGridListTileBar-title"
            >
              title
            </div>
          </div>
        </div>
      </div>
    </li>
    <li
      className="MuiGridListTile-root"
      onClick={[Function]}
      style={
        Object {
          "height": 184,
          "padding": 2,
          "width": "50%",
        }
      }
    >
      <div
        className="MuiGridListTile-tile"
      >
        <img
          alt="title2"
          src="link2.gif"
        />
        <div
          className="MuiGridListTileBar-root MuiGridListTileBar-titlePositionBottom"
        >
          <div
            className="MuiGridListTileBar-titleWrap"
          >
            <div
              className="MuiGridListTileBar-title"
            >
              title2
            </div>
          </div>
        </div>
      </div>
    </li>
    <li
      className="MuiGridListTile-root"
      onClick={[Function]}
      style={
        Object {
          "height": 184,
          "padding": 2,
          "width": "50%",
        }
      }
    >
      <div
        className="MuiGridListTile-tile"
      >
        <img
          alt="title1"
          src="link1_mp4"
        />
        <div
          className="MuiGridListTileBar-root MuiGridListTileBar-titlePositionBottom"
        >
          <div
            className="MuiGridListTileBar-titleWrap"
          >
            <div
              className="MuiGridListTileBar-title"
            >
              title1
            </div>
          </div>
        </div>
      </div>
    </li>
  </ul>
</div>
`;

exports[`GalleryGrid component Matches the snapshots for GalleryGrid with different screen width 1`] = `
<div
  className="makeStyles-root-1"
>
  <ul
    className="MuiGridList-root"
    style={
      Object {
        "margin": -2,
      }
    }
  >
    <li
      className="MuiGridListTile-root"
      onClick={[Function]}
      style={
        Object {
          "height": 184,
          "padding": 2,
          "width": "50%",
        }
      }
    >
      <div
        className="MuiGridListTile-tile"
      >
        <img
          alt="title"
          src="link"
        />
        <div
          className="MuiGridListTileBar-root MuiGridListTileBar-titlePositionBottom"
        >
          <div
            className="MuiGridListTileBar-titleWrap"
          >
            <div
              className="MuiGridListTileBar-title"
            >
              title
            </div>
          </div>
        </div>
      </div>
    </li>
  </ul>
</div>
`;

exports[`GalleryGrid component Matches the snapshots for GalleryGrid with different screen width 2`] = `
<div
  className="makeStyles-root-1"
>
  <ul
    className="MuiGridList-root"
    style={
      Object {
        "margin": -2,
      }
    }
  >
    <li
      className="MuiGridListTile-root"
      onClick={[Function]}
      style={
        Object {
          "height": 184,
          "padding": 2,
          "width": "33.333333333333336%",
        }
      }
    >
      <div
        className="MuiGridListTile-tile"
      >
        <img
          alt="title"
          src="link"
        />
        <div
          className="MuiGridListTileBar-root MuiGridListTileBar-titlePositionBottom"
        >
          <div
            className="MuiGridListTileBar-titleWrap"
          >
            <div
              className="MuiGridListTileBar-title"
            >
              title
            </div>
          </div>
        </div>
      </div>
    </li>
  </ul>
</div>
`;

exports[`GalleryGrid component Matches the snapshots for GalleryGrid with different screen width 3`] = `
<div
  className="makeStyles-root-1"
>
  <ul
    className="MuiGridList-root"
    style={
      Object {
        "margin": -10,
      }
    }
  >
    <li
      className="MuiGridListTile-root"
      onClick={[Function]}
      style={
        Object {
          "height": 200,
          "padding": 10,
          "width": "25%",
        }
      }
    >
      <div
        className="MuiGridListTile-tile"
      >
        <img
          alt="title"
          src="link"
        />
        <div
          className="MuiGridListTileBar-root MuiGridListTileBar-titlePositionBottom"
        >
          <div
            className="MuiGridListTileBar-titleWrap"
          >
            <div
              className="MuiGridListTileBar-title"
            >
              title
            </div>
          </div>
        </div>
      </div>
    </li>
  </ul>
</div>
`;

exports[`GalleryGrid component Matches the snapshots for GalleryGrid with different screen width 4`] = `
<div
  className="makeStyles-root-1"
>
  <ul
    className="MuiGridList-root"
    style={
      Object {
        "margin": -10,
      }
    }
  >
    <li
      className="MuiGridListTile-root"
      onClick={[Function]}
      style={
        Object {
          "height": 200,
          "padding": 10,
          "width": "20%",
        }
      }
    >
      <div
        className="MuiGridListTile-tile"
      >
        <img
          alt="title"
          src="link"
        />
        <div
          className="MuiGridListTileBar-root MuiGridListTileBar-titlePositionBottom"
        >
          <div
            className="MuiGridListTileBar-titleWrap"
          >
            <div
              className="MuiGridListTileBar-title"
            >
              title
            </div>
          </div>
        </div>
      </div>
    </li>
  </ul>
</div>
`;

exports[`GalleryGrid component Matches the snapshots for GalleryGrid with empty images 1`] = `null`;
